File: Forward and Inverse Kinematics Simulation/program/Inverse Kinematics.m
close all; clear; clc;
%% inital state (angle [0.5 0.3 0.8 0.1 0.2 0.5]
L_1 = 284/1000; L_2 = 225/1000; L_3 = 228.9/1000; d_6 = 55/1000;
% the position of the orgain point of link6
D_pos = [0.076483, 0.0479149, 0.688358]';
% the RPY of the orgain point of lin6
D_sta = [2.6550221, -0.9675468, -0.5583089]';
r11 = cos(D_sta(2))*cos(D_sta(3));
r12 = sin(D_sta(1))*sin(D_sta(2))*cos(D_sta(3)) - cos(D_sta(1))*sin(D_sta(3));
r13 = sin(D_sta(1))*sin(D_sta(3)) + cos(D_sta(1))*sin(D_sta(2))*cos(D_sta(3));
r21 = cos(D_sta(2))*sin(D_sta(3));
r22 = cos(D_sta(1))*cos(D_sta(3)) + sin(D_sta(1))*sin(D_sta(2))*sin(D_sta(3));
r23 = cos(D_sta(1))*sin(D_sta(2))*sin(D_sta(3)) - cos(D_sta(3))*sin(D_sta(1));
r31 = -sin(D_sta(2));
r32 = cos(D_sta(2))*sin(D_sta(1));
r33 = cos(D_sta(1))*cos(D_sta(2));
R0_7 = [r11, r12, r13; r21, r22, r23; r31, r32, r33];
T0_7 = [R0_7, D_pos; 0 0 0 1];
% *************key point************* 
T6_7 = eye(4);
T6_7(2,4) = -d_6;
T0_6 = T0_7*inv(T6_7);
R0_6 = T0_6(1:3,1:3);
dx = T0_6(1,4); dy = T0_6(2,4); dz = T0_6(3,4);

%% solve theta_1 (right)
% reflect on XOY
theta_1 = atan2(dy,dx);    % (-pi, pi]

%% solve theta_3 (right)
% in XOZ plane (needn't consider the inital angle, for example pi/2
% L_3*cos(theta_2+theta_3) - L_2*sin(theta_2) = X/cos(theta_1)  _________ 1
% L_3*sin(theta_2+theta_3) + L_2*cos(theta_2) = Z - L_1         _________ 2
% (1)^2 + (2)^2
% sin(theta_2) = ((X/cos(theta_1))^2 + (Z-L_1)^2 - (L_2)^2 -
% (L_3)^2)/(2*L_2*L_3)
sinthe_3 = (dx^2+dy^2 + (dz-L_1)^2 - (L_2)^2 - (L_3)^2)/(2*L_2*L_3);
theta_3 = asin(sinthe_3);
theta_3 = atan2(sin(theta_3),cos(theta_3)); % ******

%% solve theta_2  (right)
% L_3*(cos(theta_2)cos(theta_3)-sin(theta_2)*sin(theta_3)) -
% L_2*sin(theta_2) = X/cos(theta_1)   ___________________ 1
% L_3*(sin(theta_2)*cos(theta_3)+cos(theta_2)*sin(theta_3)) +
% L_2*cos(theta_2) = Z - L_1          ___________________ 2
A = [L_3*cos(theta_3), -(L_3*sin(theta_3)+L_2); (L_3*sin(theta_3)+L_2), L_3*cos(theta_3)];
B = [sqrt(dx^2+dy^2); dz-L_1];
% d = L_3*cos(theta_3);
% f = L_3*sin(theta_3)+L_2;
% g = D_pos(1)/cos(theta_1);
% h = D_pos(3) - L_1;
% sinthe2 = (g*f - h*d) / (d^2 - f^2);
% costhe2 = (g*d - h*f) / (d^2 - f^2);

% result = [cos;sin];
result = inv(A)*B;
theta_2 = atan2(result(2),result(1));
%theta_2 = theta_2+pi/2;
%% solve theta_4 theta_5 theta_6 
% m11 = cos(theta_1)*cos(theta_2+theta_3);
% m12 = sin(theta_1);
% m13 = cos(theta_1)*sin(theta_2+theta_3);
% m21 = sin(theta_1)*cos(theta_2+theta_3);
% m22 = -cos(theta_1);
% m23 = sin(theta_1)*sin(theta_2+theta_3);
% m31 = sin(theta_2+theta_3);
% m32 = 0;
% m33 = -cos(theta_2+theta_3);
m11 = -cos(theta_1)*sin(theta_2+theta_3);
m12 = sin(theta_1);
m13 = cos(theta_1)*cos(theta_2+theta_3);
m21 = -sin(theta_1)*sin(theta_2+theta_3);
m22 = -cos(theta_1);
m23 = sin(theta_1)*cos(theta_2+theta_3);
m31 = cos(theta_2+theta_3);
m32 = 0;
m33 = sin(theta_2+theta_3);
R0_3 = [m11, m12, m13; m21, m22, m23; m31, m32, m33];
R3_6 = (R0_3')*R0_6;
theta_4 = atan2(R3_6(2,2),R3_6(1,2))-pi;
theta_42 = theta_4+pi;
theta_5 = atan2(sqrt(R3_6(1,2)^2+R3_6(2,2)^2), R3_6(3,2))-pi/2;
theta_52 = theta_5+pi;
theta_6 = atan2(R3_6(3,3), -R3_6(3,1));
theta_62 = theta_6+pi;

%% answer
theta = [theta_1 theta_2 theta_3 theta_4 theta_5 theta_6;
    theta_1 theta_2 theta_3 theta_42 theta_5 theta_6;
    theta_1 theta_2 theta_3 theta_4 theta_52 theta_6;
    theta_1 theta_2 theta_3 theta_4 theta_5 theta_62;
    theta_1 theta_2 theta_3 theta_42 theta_52 theta_6;
    theta_1 theta_2 theta_3 theta_42 theta_5 theta_62;
    theta_1 theta_2 theta_3 theta_4 theta_52 theta_62;
    theta_1 theta_2 theta_3 theta_42 theta_52 theta_62];

File: Forward and Inverse Kinematics Simulation/program/Inverse Kinematics.py
#!/usr/bin/env python
import math
import numpy as np

def forward(angle):
    theta = angle*1
    theta[1] += math.pi/2
    theta[3] += math.pi
    theta[4] += math.pi/2
    # print(theta)
    dh = [[0,math.pi/2,0.284],
        [0.225,0,0],
        [0,math.pi/2,0],
        [0,math.pi/2,0.2289],
        [0,math.pi/2,0],
        [0,-math.pi/2,0.055]]
    x_p = np.array([0,0,0,1])
    T = np.identity(4)
    for i in range(6):
        T_tmp = np.array([math.cos(theta[i]),-math.cos(dh[i][1])*math.sin(theta[i]),
                    math.sin(dh[i][1])*math.sin(theta[i]),dh[i][0]*math.cos(theta[i]),
                    math.sin(theta[i]),math.cos(dh[i][1])*math.cos(theta[i]),
                    -math.sin(dh[i][1])*math.cos(theta[i]),dh[i][0]*math.sin(theta[i]),
                    0,math.sin(dh[i][1]),math.cos(dh[i][1]),dh[i][2],0,0,0,1]).reshape(4,4)
        T = np.dot(T,T_tmp)
    result = np.dot(T,x_p)
    # rpy
    roll = math.atan2(T[2,1],T[2,2])
    pitch = -math.asin(T[2,0])
    yaw = math.atan2(T[1,0],T[0,0])
    pose = [result[0],result[1],result[2],roll,pitch,yaw]
    return(pose)

def main():
    # angle = [0.1,0.1,0.1,0.1,0.1,0.1]
    angle = [0.5,0.3,0.8,0.1,0.2,0.5]
    # angle = [0,0,0,0,0,0]

    pose = forward(angle)
    print(pose)
    print('')

    answer = []
    l_2 = 0.225
    l_1 = 0.284
    l_3 = 0.2289
    d_6 = 0.055
    fx,fy,fz,r,p,y = pose
    T_70 = np.array([math.cos(p)*math.cos(y),math.sin(r)*math.sin(p)*math.cos(y)-math.cos(r)*math.sin(y),
                    math.sin(r)*math.sin(y)+math.cos(r)*math.sin(p)*math.cos(y),fx,
                    math.cos(p)*math.sin(y),math.cos(r)*math.cos(y)+math.sin(r)*math.sin(p)*math.sin(y),
                    math.cos(r)*math.sin(p)*math.sin(y)-math.cos(y)*math.sin(r),fy,-math.sin(p),
                    math.cos(p)*math.sin(r),math.cos(r)*math.cos(p),fz,0,0,0,1]).reshape(4,4)
    # print(T_70)
    T_76 = np.identity(4)
    T_76[1,3] -= d_6
    T_60 = np.dot(T_70,np.mat(T_76).I)
    R_60 = T_60[0:3,0:3]
    dx,dy,dz = [T_60[0,3],T_60[1,3],T_60[2,3]]
    # print(T_60)
    # print(dx,dy,dz)

    # solution
    theta_1 = math.atan2(dy,dx)
    a = -2*l_2*(dx*math.cos(theta_1)+dy*math.sin(theta_1))
    b = 2*l_2*(l_1-dz)
    c = l_3**2-((dx*math.cos(theta_1)+dy*math.sin(theta_1))**2+l_1**2-2*l_1*dz+l_2**2+dz**2)
    r = math.sqrt(a**2+b**2)
    
    for theta_2 in [math.atan2(c,math.sqrt(r**2-c**2)) - math.atan2(a,b),math.atan2(c,-math.sqrt(r**2-c**2)) - math.atan2(a,b)]:
        theta_3 = math.atan2(dx*math.cos(theta_1)+dy*math.sin(theta_1)-l_2*math.cos(theta_2),l_1+l_2*math.sin(theta_2)-dz)-theta_2

        R_30 = np.array([math.cos(theta_1)*math.cos(theta_2+theta_3),math.sin(theta_1),
                        math.cos(theta_1)*math.sin(theta_2+theta_3),math.sin(theta_1)*math.cos(theta_2+theta_3),
                        -math.cos(theta_1),math.sin(theta_1)*math.sin(theta_2+theta_3),math.sin(theta_2+theta_3),
                        0,-math.cos(theta_2+theta_3)]).reshape(3,3)

        R_63 = np.dot(np.mat(R_30).I,R_60)
        # print(R_63)

        theta_4_1 = math.atan2(R_63[1,1],R_63[0,1])
        theta_4_2 = theta_4_1 + math.pi

        theta_5_1 = math.atan2(math.sqrt(R_63[0,1]**2+R_63[1,1]**2),R_63[2,1])
        theta_5_2 = theta_5_1 + math.pi
        
        theta_6_1 = math.atan2(R_63[2,2],-R_63[2,0])
        theta_6_2 = theta_6_1 + math.pi
        
        for i in [theta_4_1,theta_4_2]:
            for j in [theta_5_1,theta_5_2]:
                for k in [theta_6_1,theta_6_2]:
                    if abs(math.cos(k)*math.sin(j)-R_63[2,0]) < 0.001 and abs(-math.cos(i)*math.sin(j)-R_63[0,1]) < 0.001:
                        theta_4 = i
                        theta_5 = j
                        theta_6 = k
                        answer.append([theta_1,theta_2-math.pi/2,theta_3,theta_4-math.pi,theta_5-math.pi/2,theta_6])

    # check and output
    for i in range(len(answer)):
        pose_inv = forward(answer[i])
        sum = 0
        for j in range(6):
            sum += (pose[j] - pose_inv[j])**2
        if sum < 0.01:
            print('angle',answer[i])
            print(pose_inv)
            print(sum)
            print('')


if __name__ == '__main__':
    main()